src/NuGet.Clients/NuGet.CommandLine/ExtensionLocator.cs

```diff
@@ -0,0 +1,118 @@
+                "NuGet",
+                "CredentialProviders");
+
+        private static readonly string CredentialProviderPattern = "CredentialProvider*.exe";
+
+        public readonly static string ExtensionsEnvar = "NUGET_EXTENSIONS_PATH";
+        public readonly static string CredentialProvidersEnvar = "NUGET_CREDENTIALPROVIDERS_PATH";
+
+        /// <summary>
+        /// Find paths to all extensions
+        /// </summary>
+        public IEnumerable<string> FindExtensions()
+        {
+            return FindAll(
+                ExtensionsDirectoryRoot,
+                ExtensionsEnvar,
+                "*.dll",
+                "*Extensions.dll"
+                );
+        }
+
+        /// <summary>
+        /// Find paths to all credential providers
+        /// </summary>
+        public IEnumerable<string> FindCredentialProviders()
+        {
+            return FindAll(
+                CredentialProvidersDirectoryRoot,
+                CredentialProvidersEnvar,
+                CredentialProviderPattern,
+                CredentialProviderPattern
+                );
+        }
+
+        /// <summary>
+        /// Helper method to locate extensions and credential providers.
+        /// The following search locations will be checked in this order:
+        /// 1) all directories under a path set by environment variable
+        /// 2) all directories under a global path, e.g. %localappdata%\nuget\commands
+        /// 3) the directory nuget.exe is located in
+        /// </summary>
+        /// <param name="globalRootDirectory">The global directory to search.  Will
+        /// also check subdirectories.</param>
+        /// <param name="customLocationEnvironmentVariable">The name of the evnrionment
+        /// variable containing a user-defined search path.  Will also check subdirectories.
+        /// The environment variable may contain a semicolon-delimited list of paths,
+        /// in which case all paths will be searched.</param>
+        /// <param name="assemblyPattern">The filename pattern to search for.</param>
+        /// <param name="nugetDirectoryAssemblyPattern">The filename pattern to search for
+        /// when looking in the nuget.exe directory. This is more restrictive so we do not
+        /// accidentally pick up NuGet dlls.</param>
+        /// <returns>An IEnumerable of paths to files matching the pattern in all searched
+        /// directories.</returns>
+        private static IEnumerable<string> FindAll(
+            string globalRootDirectory,
+            string customLocationEnvironmentVariable,
+            string assemblyPattern,
+            string nugetDirectoryAssemblyPattern)
+        {
+            IEnumerable<string> directories = new List<string>();
+
+            // Add all directories from the environment variable if available.
+            var customExtensions = Environment.GetEnvironmentVariable(customLocationEnvironmentVariable);
```

Comment:
fixed to pass in the custom paths.  The envar isn't configurable, but there is a separate one for extensions and credential providers.